module/zfs/dnode_sync.c

```diff
@@ -442,6 +443,24 @@ dnode_evict_dbufs(dnode_t *dn)
 		mutex_enter(&db->db_mtx);
 		if (db->db_state != DB_EVICTING &&
 		    zfs_refcount_is_zero(&db->db_holds)) {
+			/*
+			 * It is rare but if dn_dbufs has a dbuf with
+			 * this same bookmark but having db_state of
+			 * DB_EVICTING, we won't be able to insert
+			 * the db_marker.  Skip this dbuf if that is
+			 * the case.
+			 */
```

Comment:
I believe we can handle this case a little more efficiently.  If `dn_dbufs` has two entries for the same `db_level` and `db_blkid`, but with different `db_state`s (i.e. `DB_EVICTING` and `DB_CACHED`).  Then those entries must be adjacent in the AVL tree.  This is why the `avl_insert_here()` call fails in the debug code which verifies that the new node is inserted in the correct order.

```
VERIFY(diff > 0 ? child == 1 : child == 0) failed
PANIC at avl.c:620:avl_insert_here()
```

The most efficient way I can think to handle this is to use `AVL_PREV` to walk backwards in the tree as long as `db_level` and `db_blkid` match.  Then insert the `db_marker` at the correct location.  As I understand it that should be at most one `db` and it should be in the `DB_EVICTING` state.  Though it would be tricky to `ASSERT` that since we'd need to lock the `db` so I don't think that's necessary.  Does this make sense?

Reply:
I came up with this
```
 DUMPING DBUFS FOR OS = 118, OBJ = 0
Jul 26 14:07:02 ub16fresh kernel: [10176.850839]        dbuf 0: zb = 118:0:0:0 state = 5
Jul 26 14:07:02 ub16fresh kernel: [10176.850840]        dbuf 1: zb = 118:0:0:0 state = 4
Jul 26 14:07:02 ub16fresh kernel: [10176.850840]        dbuf 2: zb = 118:0:1:0 state = 4
Jul 26 14:07:02 ub16fresh kernel: [10176.850841]        dbuf 3: zb = 118:0:2:0 state = 4
Jul 26 14:07:02 ub16fresh kernel: [10176.850841]        dbuf 4: zb = 118:0:3:0 state = 4
Jul 26 14:07:02 ub16fresh kernel: [10176.850842]        dbuf 5: zb = 118:0:4:0 state = 4
Jul 26 14:07:02 ub16fresh kernel: [10176.850843]        dbuf 6: zb = 118:0:5:0 state = 4
Jul 26 14:07:02 ub16fresh kernel: [10176.850843] ----------------------------------------
Jul 26 14:07:02 ub16fresh kernel: [10176.850844] NOTICE: AVL_PREV 0 0 matches marker
Jul 26 14:07:02 ub16fresh kernel: [10176.850845] NOTICE: inserting db_marker at head of list


void
dnode_evict_dbufs(dnode_t *dn)
{
        dmu_buf_impl_t *db_marker, *insert_point;
        dmu_buf_impl_t *db, *db_next;

        db_marker = kmem_alloc(sizeof (dmu_buf_impl_t), KM_SLEEP);

        mutex_enter(&dn->dn_dbufs_mtx);
        dump_dnode_dbufs(dn);
        for (db = avl_first(&dn->dn_dbufs); db != NULL; db = db_next) {

#ifdef  DEBUG
                DB_DNODE_ENTER(db);
                ASSERT3P(DB_DNODE(db), ==, dn);
                DB_DNODE_EXIT(db);
#endif  /* DEBUG */

                mutex_enter(&db->db_mtx);
                if (db->db_state != DB_EVICTING &&
                    zfs_refcount_is_zero(&db->db_holds)) {
                        /*
                         * It is rare but if dn_dbufs has a dbuf with
                         * this same bookmark but having db_state of
                         * DB_EVICTING, we won't be able to insert
                         * the db_marker.  Work backward to find the correct
                         * insertion point.
                         */
                        boolean_t b = B_FALSE;
                        db_marker->db_level = db->db_level;
                        db_marker->db_blkid = db->db_blkid;
                        db_marker->db_state = DB_SEARCH;
                        insert_point = AVL_PREV(&dn->dn_dbufs, db);
                        while (insert_point != NULL  &&
                            insert_point->db_level == db->db_level &&
                            insert_point->db_blkid == db->db_blkid) {
                                cmn_err(CE_NOTE, "AVL_PREV %d %lld matches marker",
                                    insert_point->db_level, (longlong_t)insert_point->db_blkid);
                                insert_point = AVL_PREV(&dn->dn_dbufs,
                                    insert_point);
                                b = B_TRUE;
                        }
                        if (insert_point == NULL) {
                                if (b) cmn_err(CE_NOTE, "inserting db_marker at head of list");
                                avl_insert_here(&dn->dn_dbufs, db_marker,
                                    avl_first(&dn->dn_dbufs), AVL_BEFORE);
                        }
                        else {
                                if (b) cmn_err(CE_NOTE, "inserting db_marker after %d %lld",
                                    insert_point->db_level, (longlong_t)insert_point->db_blkid);
                                avl_insert_here(&dn->dn_dbufs, db_marker,
                                    insert_point, AVL_AFTER);
                        }
```
It tests out fine, though in this unit test it seems that when it happens, the DB_EVICTING db is always first in the tree, and hits the "insert_point == NULL" case.

Also I figured out the dsl_dataset_snapshot_check error.

Reply:
Great, that basic logic makes sense to me.  As does it always hitting the "insert_point == NULL" case, since I believe we're only going to see this for blkid 0 for this test case.

Reply:
Ok I'll clean it up and get a CI run going.  Thanks.